File: permissionsex-core/src/main/java/ninja/leaping/permissionsex/util/command/ChildCommands.java
/**
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.leaping.permissionsex.util.command;

import com.google.common.base.Optional;
import com.google.common.base.Predicate;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Iterables;
import ninja.leaping.permissionsex.util.command.args.ArgumentParseException;
import ninja.leaping.permissionsex.util.command.args.CommandArgs;
import ninja.leaping.permissionsex.util.command.args.CommandElement;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static ninja.leaping.permissionsex.util.Translations._;
import static ninja.leaping.permissionsex.util.Translations.untr;

/**
 * Utility methods for handling child commands
 */
public class ChildCommands {
    private ChildCommands() {}

    public static CommandElement args(CommandSpec... children) {
        Map<String, CommandSpec> mapping = new HashMap<>();
        for (CommandSpec child : children) {
            List<String> aliases = child.getAliases();
            if (aliases.isEmpty()) {
                continue; // Unnamable command -- TODO maybe warn?
            }
            final String primaryName = aliases.get(0);
            if (mapping.containsKey(primaryName)) {
                continue; // oh well, we're presented with an ordered collection so hopefully whoever is calling us knows what they're doing
            }
            mapping.put(primaryName, child);
        }

        for (CommandSpec child : children) {
            List<String> aliases = child.getAliases();
            for (int i = 1; i < aliases.size(); ++i) {
                if (!mapping.containsKey(aliases.get(i))) {
                    mapping.put(aliases.get(i), child);
                }
            }
        }
        return new ChildCommandElement(mapping);
    }

    private static class ChildCommandElement extends CommandElement {
        private static final AtomicInteger COUNTER = new AtomicInteger();
        private final Map<String, CommandSpec> children;

        private ChildCommandElement(Map<String, CommandSpec> children) {
            super(untr("child" + COUNTER.getAndIncrement()));
            this.children = ImmutableMap.copyOf(children);
        }

        @Override
        public void parse(CommandArgs args, CommandContext context) throws ArgumentParseException {
            super.parse(args, context);
            CommandSpec spec = context.getOne(getKey().getUntranslated());
            spec.parse(args, context);
        }

        @Override
        protected Object parseValue(CommandArgs args) throws ArgumentParseException {
            final String key = args.next();
            if (!children.containsKey(key.toLowerCase())) {
                throw args.createError(_("Input command %s was not a valid subcommand!", key));
            }

            return children.get(key.toLowerCase());
        }

        @Override
        public <TextType> List<String> tabComplete(final Commander<TextType> src, CommandArgs args, CommandContext context) {
            final Optional<String> commandComponent = args.nextIfPresent();
                if (commandComponent.isPresent()) {
                    if (args.hasNext()) {
                        CommandSpec child = children.get(commandComponent.get());
                        if (child != null) {
                            try {
                                child.checkPermission(src);
                                return child.tabComplete(src, args, context); // todo make this more correct
                            } catch (CommandException e) {
                            }
                        }
                        return ImmutableList.of();
                    } else {
                        return ImmutableList.copyOf(Iterables.filter(filterCommands(src), new Predicate<String>() {
                            @Override
                            public boolean apply(String input) {
                                return input.startsWith(commandComponent.get());
                            }
                        }));
                    }
                } else {
                    return ImmutableList.copyOf(children.keySet());
                }
        }

        @Override
        public <TextType> TextType getUsage(Commander<TextType> context) {
            List<Object> args = new ArrayList<>(Math.max(0, children.size() * 2 - 1));
            Iterable<String> filteredCommands = Iterables.filter(filterCommands(context), new Predicate<String>() {
                @Override
                public boolean apply(String input) {
                    return children.get(input).getAliases().get(0).equals(input); // Restrict to primary aliases in usage
                }
            });

            for (Iterator<String> it = filteredCommands.iterator(); it.hasNext();) {
                args.add(it.next());
                if (it.hasNext()) {
                    args.add("|");
                }
            }
            return context.fmt().combined(args.toArray());
        }

        private Iterable<String> filterCommands(final Commander<?> src) {
            return Iterables.filter(children.keySet(), new Predicate<String>() {
                @Override
                public boolean apply(String input) {
                    CommandSpec child = children.get(input);
                    try {
                        child.checkPermission(src);
                        return true;
                    } catch (CommandException ex) {
                        return false;
                    }
                }
            });
        }
    }

    public static CommandExecutor executor(CommandElement arg) {
        return new ChildCommandExecutor(arg.getKey().getUntranslated(), null);
    }

    public static CommandExecutor optionalExecutor(CommandElement arg, CommandExecutor fallbackExecutor) {
        return new ChildCommandExecutor(arg.getKey().getUntranslated(), fallbackExecutor);
    }

    private static class ChildCommandExecutor implements CommandExecutor {
        private final String key;
        private final CommandExecutor fallbackExecutor;

        private ChildCommandExecutor(String key, CommandExecutor fallbackExecutor) {
            this.key = key;
            this.fallbackExecutor = fallbackExecutor;
        }

        @Override
        public <TextType> void execute(Commander<TextType> src, CommandContext args) throws CommandException {
            CommandSpec spec = args.getOne(key);
            if (spec == null) {
                if (fallbackExecutor != null) {
                    fallbackExecutor.execute(src, args);
                    return;
                } else {
                    throw new CommandException(_("Invalid subcommand state -- only one command spec must be provided for child arg %s", key));
                }
            }
            spec.checkPermission(src);
            spec.getExecutor().execute(src, args);
        }
    }
}

File: permissionsex-core/src/main/kotlin/ca/stellardrift/permissionsex/commands/parse/ChildCommands.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.commands.parse

import ca.stellardrift.permissionsex.commands.commander.Commander
import ca.stellardrift.permissionsex.commands.parse.StructuralArguments.optional
import ca.stellardrift.permissionsex.commands.CommonMessages.ERROR_CHILDREN_STATE
import ca.stellardrift.permissionsex.commands.CommonMessages.ERROR_CHILDREN_UNKNOWN
import ca.stellardrift.permissionsex.util.join
import ca.stellardrift.permissionsex.util.unaryPlus
import net.kyori.text.Component
import java.util.Locale
import java.util.concurrent.atomic.AtomicInteger

/**
 * Build the children for a parent command
 */
@CommandDsl
class ChildCommandBuilder {
    internal val children: MutableList<CommandSpec> = mutableListOf()
    private var fallback: CommandAction? = null

    /**
     * Add the already existing command as a child of this one
     */
    fun child(spec: CommandSpec) {
        children.add(spec)
    }

    /**
     * Create a new command that will be added as a child
     */
    fun child(vararg aliases: String, init: CommandBuilder.() -> Unit) {
        val builder = CommandBuilder(aliases.toList())
        builder.init()
        this.children.add(builder.build())
    }

    /**
     * Set an action that will be run if no child command is specified
     *
     * Declaring a fallback makes the child command optional.
     */
    fun fallback(action: CommandAction) {
        this.fallback = action
    }

    /**
     * Build a
     */
    internal fun build(): Triple<Component, CommandElement, CommandAction> {
        val element = buildElement()
        val returnedElement = if (fallback != null) {
            optional(element)
        } else {
            element
        }
        return Triple(element.key, returnedElement, buildAction(element))
    }

    private fun buildElement(): ChildCommandElement {
        val mapping: MutableMap<String, CommandSpec> = mutableMapOf()
        for (child in children) {
            val aliases = child.aliases
            if (aliases.isEmpty()) {
                continue  // Unnamable command -- TODO maybe warn?
            }
            mapping.putIfAbsent(aliases[0], child)
        }
        for (child in children) {
            for (alias in child.aliases) {
                mapping.putIfAbsent(alias, child)
            }
        }
        return ChildCommandElement(mapping)
    }

    private fun buildAction(arg: ChildCommandElement): CommandAction {
        val fallback = this.fallback
        val key = arg.key.toContextKey()
        return outer@ { src, args ->
            val spec =
                args.getOne<CommandSpec>(key)
                    ?: if (fallback != null) {
                        fallback(src, args)
                        return@outer
                    } else {
                        throw CommandException(ERROR_CHILDREN_STATE(key))
                    }
            spec.checkPermission(src)
            spec.executor?.invoke(src, args)
        }

    }
}

class ChildCommandElement internal constructor(private val children: Map<String, CommandSpec>) : CommandElement() {
    val key = +"child${COUNTER.getAndIncrement()}"

    @Throws(ArgumentParseException::class)
    override fun parse(
        args: CommandArgs,
        context: CommandContext
    ) {
        val key = args.next().toLowerCase(Locale.ROOT)
        val spec = children[key] ?: throw args.createError(ERROR_CHILDREN_UNKNOWN(key))
        context.putArg(this.key, spec)
        spec.parse(args, context)
    }

    override fun tabComplete(src: Commander, args: CommandArgs, context: CommandContext): List<String> {
        val commandComponent = args.nextIfPresent()
        return if (commandComponent != null) {
            if (args.hasNext()) {
                val child = children[commandComponent]
                if (child != null) {
                    try {
                        child.checkPermission(src)
                        return child.tabComplete(src, args, context) // todo make this more correct
                    } catch (e: CommandException) {
                    }
                }
                listOf()
            } else {
                filterCommands(src).filter { it.startsWith(commandComponent, ignoreCase = true) }.toList()
            }
        } else {
            children.keys.toList()
        }
    }

    override fun getUsage(src: Commander): Component {
        return filterCommands(src)
            .filter { children[it]?.aliases?.get(0) == it }
            .map { +it}
            .join(PIPE)
    }

    private fun filterCommands(src: Commander): Sequence<String> {
        return children.asSequence()
            .filter { (_, v) -> v.hasPermission(src) }
            .map { (k, _) -> k }
    }

    companion object {
        private val COUNTER = AtomicInteger()
    }
}

